### PR TITLE
💥 Remove the legacy circuit surface for v4

### LINKED ISSUE #2085: 🧹 MQT Core v4 cleanup and ownership tracking
🤖 *AI text below* 🤖

## Goal

Use the MQT Core v4 boundary to reduce legacy surface, remove unused APIs, and move specialized functionality to the repositories that actually own it. The aim is a smaller and clearer Core package with fewer public targets, transitive dependencies, compatibility layers, and domain-specific concepts.

The sub-issues are grouped into four tracks:

1. straightforward removals and API pruning;
2. dependency and packaging simplification;
3. ownership transfers to QMAP, QCEC, DDSIM, and QDMI-on-IQM; and
4. bounded architectural cleanup of CoreIR, DD, QDMI, QIR, and OpenQASM support.

Each sub-issue should remain independently reviewable. Cross-repository transfers should prepare the downstream owner before removing the corresponding Core API.

## Scope principles

- Backward compatibility is not a reason to retain deprecated or unused machinery in v4.
- Preserve useful user-facing behavior while simplifying its implementation and ownership.
- Do not replace one maintenance burden with a generic abstraction that has no demonstrated consumer.
- Record removed targets, headers, namespaces, Python APIs, options, and migration paths in `CHANGELOG.md` and `UPGRADING.md`.
- Validate installed shared and static CMake packages, Python wheels, downstream source builds, documentation, and dependency inspection where applicable.

## Explicit non-goals

- The compiler collection remains in MQT Core.
- The generic QDMI layer remains in MQT Core for now.
- The Qiskit and PennyLane adapters are not being assigned a new owner before v4.
- The DDSIM and SC QDMI devices remain supported.
- Jeff remains supported.
- This tracker does not pursue CMake templating merely to remove repeated declarations.

## Audit basis

The initial inventory was performed against MQT Core `main` at `2953866e3f0b424c434cd4c4f882795e3a18e85e`. Individual issues should refresh usage and downstream evidence before implementation.

### LINKED ISSUE #2097: ♻️ Consolidate OpenQASM import around the compiler frontend
🤖 *AI text below* 🤖

## Goal

Reduce the duplicated OpenQASM parser and semantic implementation while preserving lightweight `QuantumComputation` consumers.

## Direction to evaluate

- Make the compiler's OpenQASM parser and semantic analysis the canonical frontend.
- Provide separate emitters for MLIR programs and `QuantumComputation` rather than requiring every CoreIR consumer to link the complete LLVM/MLIR stack.
- Retire the legacy parser only after behavioral parity is demonstrated.
- Coordinate with the compiler translation work tracked in #1733.

## Required compatibility

- OpenQASM 2 and supported OpenQASM 3 input.
- Python `QuantumComputation` imports.
- DDSIM inputs and layout comments.
- Symbolic parameters, custom gates, control flow, diagnostics, and source locations to the extent supported today.

## Acceptance criteria

- One lexer/parser/semantic implementation serves both compiler and classic IR paths.
- Lightweight CoreIR consumers do not acquire a runtime LLVM/MLIR dependency merely to parse OpenQASM.
- Differential tests cover accepted programs, rejected programs, diagnostics, and constructed circuit semantics before the old frontend is deleted.

### LINKED ISSUE #2103: ♻️ Consolidate duplicated DD circuit simulation
🤖 *AI text below* 🤖

## Goal

Eliminate duplicated circuit-execution logic between Core's `dd::Simulation` helpers and DDSIM's `CircuitSimulator` without introducing a cyclic or inverted dependency.

## Constraints

- The DDSIM QDMI reference device remains in MQT Core for v4 and requires circuit execution.
- Core currently exposes Python `sample`, `simulate`, and `simulate_statevector` helpers.
- SyReC and QCEC have identified dependencies on parts of the current API.

## Work

- Inventory the exact shared execution kernel: unitary application, virtual swaps, measurement sampling, dynamic shots, reset, classical control, RNG, permutations, and garbage handling.
- Decide which minimal primitives stay in Core for the QDMI device and which public simulator API belongs in DDSIM.
- Remove unused virtual-operation helpers where possible.
- Migrate Python and downstream consumers explicitly.

## Acceptance criteria

- One implementation owns each execution behavior.
- Core does not depend on DDSIM merely to build its retained QDMI device.
- DDSIM remains the owner of the public simulator abstraction.
- Dynamic circuits, repeated jobs, seeding, output permutations, and Python results retain coverage.

### PR DESCRIPTION
**Pull request by simon1hofmann**
🤖 *AI text below* 🤖

## Description

Remove the complete legacy circuit surface for MQT Core v4: `qc::QuantumComputation`, CoreIR/CoreQASM targets and headers, `mqt.core.ir`, `mqt.core.load`, classic Qiskit converters, and circuit-oriented DD adapters. MQT Core v4 deliberately provides no compatibility aliases, legacy circuit constructs, or density-simulation layer.

CoreDD retains only its DD-native gate, control, qubit, and permutation vocabulary. The compiler-backed DDSIM QDMI provider accepts OpenQASM through QC/QCO and QIR. Terminal sampling simulates once and draws every shot from the resulting state; statevector extraction supports programs with only terminal measurements or no measurements. Expected compiler and simulation failures become failed jobs without exception-driven control flow.

When MLIR is disabled, MQT Core does not build the DDSIM device because no execution frontend remains. CoreDD, the QDMI driver, the superconducting device, and their applicable tests remain available.

This PR is the second and final layer of #2079 → #2288. The stack keeps the small math execution extensions and intentionally does not include density simulation. Symbolic Python execution can follow in a later release.

The local C++ lint wrapper now prints Clang diagnostics as file annotations instead of reporting only their count.

This implementation and PR description were produced with GPT-5.6 via Codex under explicit maintainer authorization. Maintainer review remains required.

## Validation

- `uvx nox -s lint`
- `uvx nox -s cpp-lint -- 5ebdb6da5da533b40ccdf19a5ab6af4e0a294d49` (Clang 22; zero findings)
- `uvx nox -s tests-3.13` (745 passed, one Python 3.14-only test skipped)
- `cmake --build --preset debug -j2`
- `ctest --preset debug --output-on-failure -j2` (3,755 passed)
- `cmake --build --preset debug-no-mlir -j2`
- `ctest --preset debug-no-mlir --output-on-failure -j2` (511 passed, one intentional device-specification skip)
- The no-MLIR Ninja target graph contains no DDSIM device target.
- `git diff --check`

Fixes #2097
Fixes #2103

## Checklist

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [ ] The changes are fully tested and pass the CI checks.
- [ ] I have reviewed my own code changes.

**If PR contains AI-assisted content:**

- [x] Any agent that created, edited, or submitted GitHub content was explicitly authorized for that scope, as required by our [AI Usage Guidelines](https://github.com/munich-quantum-toolkit/core/blob/main/docs/ai_usage.md).
- [x] Every agent-authored or agent-edited public text body begins with the visible disclosure `🤖 *AI text below* 🤖` (titles are exempt).
- [x] I have disclosed AI assistance in the PR description.
- [ ] I confirm that I have personally reviewed and understood all AI-generated content, and accept full responsibility for it.

### CODE REVIEW

File: include/mqt-core/dd/GateMatrixDefinitions.hpp
Comment (burgholzer) [inline include/mqt-core/dd/GateMatrixDefinitions.hpp:1]: Could we somehow eliminate the redundancy here and derive these somehow from the matrices we compute in the MLIR subtree?

File: mlir/include/mlir/Dialect/QIR/Execution/Runtime/Runtime.h
Comment (burgholzer) [inline mlir/include/mlir/Dialect/QIR/Execution/Runtime/Runtime.h:1]: This is one of the files that would likely also need adjustments if we were to drop the duplicate gate matrix tracking.

File: python/mqt/core/plugins/qiskit/backend.py [hunk 1]
```diff
@@ -123,6 +123,20 @@ def _serialize_to_qasm3(circuit: QuantumCircuit, backend: QDMIBackend) -> str:
...
+    if circuit.num_clbits:
+        initialization = circuit.copy_empty_like(vars_mode="captures")
+        initialization.global_phase = 0
+        for clbit in initialization.clbits:
+            initialization.store(
+                clbit,
+                False,  # ruff: ignore[boolean-positional-value-in-call] Qiskit store arguments are positional-only.
+            )
+        circuit = circuit.copy()
+        circuit.compose(initialization, front=True, inplace=True, inline_captures=True)
```
Comment (burgholzer) [inline python/mqt/core/plugins/qiskit/backend.py]: This feels a bit like a hack. Can we find a better way to work around this?

File: src/qdmi/devices/dd/CMakeLists.txt [hunk 2]
```diff
@@ -38,17 +38,15 @@ if(NOT TARGET ${TARGET_NAME})
...
-    endif()
-    target_link_libraries(${TARGET_NAME} PRIVATE MQT::CoreQIRJIT MQT::CoreQIRRuntime)
+  target_link_libraries(${TARGET_NAME} PRIVATE MQT::CoreDD MQT::CoreQDMICommon)
+  mqt_llvm_target_disable_rtti(${TARGET_NAME})
+  if(UNIX AND NOT APPLE)
+    # Keep symbols from static dependencies local. A plugin-enabled host can otherwise interpose
+    # LLVM's process-wide registries when it loads this device.
+    target_link_options(${TARGET_NAME} PRIVATE "LINKER:--exclude-libs,ALL")
   endif()
+  target_link_libraries(${TARGET_NAME} PRIVATE MLIRQCODDFunctionality MQTCompilerPipeline
```
Comment (burgholzer) [inline src/qdmi/devices/dd/CMakeLists.txt:48]: It feels a bit weird that this links the Compiler Pipeline. Is this really on purpose or should we be defining a separate target for a certain concept?

File: src/qdmi/devices/dd/Device.cpp [hunk 3]
```diff
@@ -150,14 +151,61 @@ constexpr auto OPERATION_ADDRESSES = makeOperationAddresses(OPERATIONS);
...
+    key.reserve(outcome.size());
+    auto offset = outcome.size();
+    for (const auto width : widths | std::views::reverse) {
+      offset -= width;
+      key.append(outcome, offset, width);
+    }
+    reordered.emplace(std::move(key), count);
+  }
+  return reordered;
+}
```
Comment (burgholzer) [inline src/qdmi/devices/dd/Device.cpp]: Would this be a good time to try and avoid having to constantly reorder? I assume we could just adjust the semantics of the simulator to match the expectations.

File: src/qdmi/devices/dd/Device.cpp [hunk 4]
```diff
@@ -470,23 +517,49 @@ auto MQT_DDSIM_QDMI_Device_Job_impl_d::submitQASMProgramSampling()
...
+      return false;
+    }
+    const auto entryPoint = mlir::mqt::getEntryPoint(qcoProgram->module());
+    if (!entryPoint) {
+      std::cerr << "Error: QCO program has no entry point\n";
+      return false;
+    }
+    dd::Package package;
+    auto rng = makeJobRng(seed_);
+    auto counts = mlir::qco::sample(entryPoint, package, numShots_, rng);
```
Comment (burgholzer) [inline src/qdmi/devices/dd/Device.cpp]: Just noticed this here, but the sampling function does not actually need to receive a DD package from external. As it only returns counts, it can construct the DD package internally. That should generally simplify the signature a little bit.
I am also wondering whether the RNG logic should be internalized in the respective function so that people can just pass a seed to these functions. Would also simplify some code.

File: test/python/qdmi/test_qdmi.py [hunk 5]
```diff
@@ -768,6 +768,22 @@ def simulator_job(ddsim_device: Device) -> Job:
...
+    sample_job = ddsim_device.submit_job(program, ProgramFormat.QASM3, num_shots=4)
+    sample_job.wait()
+    assert sample_job.get_counts() == {"": 4}
+
+    state_job = ddsim_device.submit_job(program, ProgramFormat.QASM3, num_shots=0)
+    state_job.wait()
+    assert state_job.get_dense_statevector() == [1 + 0j]
+    assert state_job.get_dense_probabilities() == [1.0]
+    assert state_job.get_sparse_statevector() == {"": 1 + 0j}
+    assert state_job.get_sparse_probabilities() == {"": 1.0}
```
Comment (burgholzer) [inline test/python/qdmi/test_qdmi.py]: I would be tempted to argue that these should be empty dicts and lists for zero-qubit programs. feels a bit cleaner as corner case handling.

File: test/python/test_python_surface.py
Comment (burgholzer) [inline test/python/test_python_surface.py:1]: This feels unnecessary and may be removed

File: test/qdmi/driver/test_driver.cpp [hunk 6]
```diff
@@ -57,7 +57,7 @@ MATCHER_P2(IsBetween, a, b,
 } // namespace
 } // namespace testing
 
-namespace qc {
+namespace {
```
Comment (burgholzer) [inline test/qdmi/driver/test_driver.cpp:60]: Double namespace here.

File: CHANGELOG.md [hunk 7]
```diff
@@ -120,13 +116,14 @@ releases may include breaking changes.
 
 ### Removed
 
+- 💥 Remove the classic circuit APIs and adapters. Use compiler-backed QC/QCO
+  APIs in v4 or MQT Core 3.9.x for legacy code ([#2288]) ([**@simon1hofmann**])
```
Comment (burgholzer) [inline CHANGELOG.md]: This should likely mention `qc::QuantumComputation` explicitly (likely also `MQT::CoreIR`.

File: CHANGELOG.md [hunk 8]
```diff
@@ -108,8 +106,6 @@ releases may include breaking changes.
   ([**@burgholzer**])
 - 🐛 Protect process-wide DD, IR, and QDMI state for free-threaded Python
   ([#2209]) ([**@burgholzer**])
-- 💥 Prune dead and misleading CoreIR APIs and remove random-number generator
-  state from `QuantumComputation` ([#2111], [#2112]) ([**@simon1hofmann**])
```
Comment (burgholzer) [inline CHANGELOG.md:113]: I think these PRs may still be added to the new changelog entry.

File: pyproject.toml [hunk 9]
```diff
@@ -290,7 +290,7 @@ select = ["cp*-win_arm64", "cp315*"]
 before-test = ""
 test-groups = []
 test-sources = []
-test-command = "python -c \"import mqt.core.dd, mqt.core.ir, mqt.core.mlir, mqt.core.qdmi\""
+test-command = "python -c \"import importlib.util; import mqt.core, mqt.core.dd, mqt.core.mlir, mqt.core.qdmi; assert not hasattr(mqt.core, 'load'); assert importlib.util.find_spec('mqt.core.ir') is None\""
```
Comment (burgholzer) [inline pyproject.toml]: Overkill, just drop `ir`

File: README.md [hunk 10]
```diff
@@ -97,14 +97,18 @@ uv pip install mqt.core
...
+program = compile_program("""OPENQASM 3.0;
+include "stdgates.inc";
+qubit[2] q;
+bit[2] result;
+h q[0];
+cx q[0], q[1];
+result = measure q;
+""")
+
+print(program.ir)
```
Comment (burgholzer) [inline README.md]: It likely makes sense to show an end-to-end demonstration here (compiling to QIR and executing via QDMI through the DDSIM device)

File: UPGRADING.md [hunk 11]
```diff
@@ -6,6 +6,37 @@ of changes including minor and patch releases, please refer to the
...
+qc_program = QCProgram.from_qasm_file("circuit.qasm")
+qco_program = compile_program(qc_program, output=OutputFormat.QCO_OPTIMIZED)
+```
+
+Use `QCProgram.from_qasm_str` for source text, `QCProgram.from_qiskit` for a
+Qiskit `QuantumCircuit`, and `QCProgram.to_qiskit` for conversion back to
+Qiskit. For decision-diagram simulation, lower to a `QCOProgram` and use its
+`sample`, `simulate`, or `build_functionality` methods. The circuit-taking
+functions in `mqt.core.dd` and the operation-taking `DDPackage` methods have
+been removed; the raw vector and matrix DD constructors remain available.
```
Comment (burgholzer) [inline UPGRADING.md:32]: Now that I read this, it likely makes sense to introduce convenience functions for simulating/sampling/building QCPrograms.
One may even think about providing an umbrella function for each of them that uses a union of all supported program formats and internally uses the most direct way to go to QCO and call the actual simulation routine.
This should make general usage of the simulation capabilities quite convenient.

File: UPGRADING.md [hunk 12]
```diff
@@ -6,6 +6,37 @@ of changes including minor and patch releases, please refer to the
...
+targets, and the installed `ir/` and `qasm3/` headers. Python no longer provides
+`mqt.core.ir`, the top-level `mqt.core.load` helper, or the classic
+`mqt.core.plugins.qiskit.mqt_to_qiskit` and `qiskit_to_mqt` converters. MQT Core
+4 deliberately provides no compatibility alias or adapter for these APIs.
+
+Use the compiler-backed Python interface in new code:
+
+```python
+from mqt.core.mlir import OutputFormat, QCProgram, compile_program
+
```
Comment (burgholzer) [inline UPGRADING.md:22]: This should likely contain a reference to MQT Core v3.x and that that release series keeps the respective libraries and that consumers depending on it shall use that major version series.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Alright, I reviewed this in quite a bit of detailed and found some further potential improvements. @simon1hofmann can I leave these to you?
(It seems like I also broke something in the last iteration)

File: src/qdmi/devices/dd/Device.cpp (hunk 4 above)
Comment (simon1hofmann) [reply src/qdmi/devices/dd/Device.cpp]: Replaced the public zero-state sampling signature with shots plus a seed. Function now owns the package/RNG internally.

File: include/mqt-core/dd/GateMatrixDefinitions.hpp
Comment (simon1hofmann) [reply include/mqt-core/dd/GateMatrixDefinitions.hpp:1]: Not sure if CoreDD can derive from optional MLIR, would probably need a backend-neutral matrix layer.

File: src/qdmi/devices/dd/CMakeLists.txt (hunk 2 above)
Comment (simon1hofmann) [reply src/qdmi/devices/dd/CMakeLists.txt:48]: [Device.cpp:159–169](https://github.com/munich-quantum-toolkit/core/pull/2288/changes#diff-36511f74588a66ee354fd0caab0d8cb84e2f29786acf0845b4aeb3d8f5513dcdR159-R169) uses APIs implemented in Programs.cpp.

File: UPGRADING.md (hunk 12 above)
Comment (simon1hofmann) [reply UPGRADING.md:22]: Explained here: https://github.com/munich-quantum-toolkit/core/pull/2288/changes#diff-f411a5009131db5a6c7242b49ca78488f2d274ef9e5e5f4de9262a42108a5386R34-R38

MQT Core 3.9.x continues to provide the classic circuit interfaces for
repositories that have not migrated. Pin `mqt-core~=3.9.0` and stay on the 3.9
release line for those consumers. C++ consumers should likewise use the 3.9
release branch or a matching 3.9 version constraint. MQT Core 3.9 and 4 cannot
provide their Python or CMake packages in the same environment.

File: UPGRADING.md (hunk 11 above)
Comment (simon1hofmann) [reply UPGRADING.md:32]: Should we do that as part of this PR or in a follow-up for another minor release?

File: mlir/include/mlir/Dialect/QIR/Execution/Runtime/Runtime.h
Comment (simon1hofmann) [reply mlir/include/mlir/Dialect/QIR/Execution/Runtime/Runtime.h:1]: Relates to this comment: https://github.com/munich-quantum-toolkit/core/pull/2288#discussion_r3913456167

File: src/qdmi/devices/dd/Device.cpp (hunk 3 above)
Comment (simon1hofmann) [reply src/qdmi/devices/dd/Device.cpp]: Yes, agreed. I moved the ordering into QCO’s sampling semantics: returned CBit registers now use conventional count-string order (last returned register first, with each register MSB-first). The DDSIM device can therefore consume the histogram directly, and reverseRegisterSegments has been removed.

File: python/mqt/core/plugins/qiskit/backend.py (hunk 1 above)
Comment (simon1hofmann) [reply python/mqt/core/plugins/qiskit/backend.py]: 🤖 This response was generated with AI assistance.
I traced this through the complete path. Qiskit starts classical memory at zero, whereas OpenQASM 3 deliberately leaves uninitialized classical declarations undefined. Changing the OpenQASM frontend or simulator would therefore change the semantics of arbitrary OpenQASM programs.
Qiskit’s OpenQASM 3 exporter currently has no public mechanism for initializing the generated declarations, so materializing Qiskit’s implicit initialization as explicit stores at the serializer boundary is the narrowest semantics-preserving solution. I simplified the initialization layer to drop unused real-time variables and compose it non-mutatively. A register-wide store is not currently viable because Qiskit emits c = 0, which is not a bit-register value accepted by our OpenQASM frontend. Longer term, this can be removed if Qiskit adds support for initialized classical declarations.

Comment (simon1hofmann) [reply python/mqt/core/plugins/qiskit/backend.py]: Simplified it a little bit in https://github.com/munich-quantum-toolkit/core/commit/8c65b0fe5

Comment (burgholzer) [reply python/mqt/core/plugins/qiskit/backend.py]: > A register-wide store is not currently viable because Qiskit emits c = 0, which is not a bit-register value accepted by our OpenQASM frontend. Longer term, this can be removed if Qiskit adds support for initialized classical declarations.

Wouldn't it be cleaner to simply add support for this assignment statement (which is valid OpenQASM 3 IIRC) and to clean this up directly here?

File: src/qdmi/devices/dd/CMakeLists.txt (hunk 2 above)
Comment (burgholzer) [reply src/qdmi/devices/dd/CMakeLists.txt:48]: Hm. Maybe the `Programs` module should be its own library that can be linked against. It conceptionally does not make too much sense to be that the QDMI device would include parts of the compiler pipeline. (Maybe some of the other changes below which I haven't looked into in detail yet already triggered some further changes here)

File: src/qdmi/devices/dd/Device.cpp (hunk 3 above)
Comment (burgholzer) [reply src/qdmi/devices/dd/Device.cpp]: Nice! that feels like a genuinely useful simplification! 🙌🏼

File: UPGRADING.md (hunk 11 above)
Comment (burgholzer) [reply UPGRADING.md:32]: Hm. It kind of feels like something that could genuinely be useful and that now would be the right time to provide that convenient API in the initial launch.
Right now this feels like it is exposing too many of the internals to be convenient to use.
I would lean towards getting this into the v4.0 release. But maybe in a separate follow-up to this PR here to not overcomplicate it and hold it up for too long.
The large removal here will also save large amounts of CI resources, so we should try to get it in quickly.

File: UPGRADING.md (hunk 12 above)
Comment (burgholzer) [reply UPGRADING.md:22]: This feels a bit too limited to 3.9 for me.
We are about to release 3.10 in conjunction with v4.0 (I suppose), so this would already be outdated by then.
Should we maybe refer to this more broadly as the MQT Core v3 (release) series?

File: src/qdmi/devices/dd/CMakeLists.txt (hunk 2 above)
Comment (simon1hofmann) [reply src/qdmi/devices/dd/CMakeLists.txt:48]: 🤖 This response was generated with AI assistance.

Agreed. A mechanical split of `Programs.cpp` is not sufficient because it also implements compiler transformations, target compilation, and default-pipeline orchestration. I kept that architectural change out of this PR and opened #2329 to establish the appropriate library boundary.

File: UPGRADING.md (hunk 11 above)
Comment (simon1hofmann) [reply UPGRADING.md:32]: 🤖 This response was generated with AI assistance.

Agreed. I kept #2288 focused on the removal and opened #2330 for the convenience APIs so they can still be completed for v4 without holding up this PR.

File: UPGRADING.md (hunk 12 above)
Comment (simon1hofmann) [reply UPGRADING.md:22]: 🤖 This response was generated with AI assistance.

Updated `UPGRADING.md` to refer to the complete MQT Core v3 release series and recommend `mqt-core>=3,<4`; the changelog now uses the same wording.

File: src/qdmi/devices/dd/CMakeLists.txt (hunk 2 above)
Comment (burgholzer) [reply src/qdmi/devices/dd/CMakeLists.txt:48]: That makes sense 👍🏼 Let's resolve this here

File: include/mqt-core/dd/GateMatrixDefinitions.hpp
Comment (burgholzer) [reply include/mqt-core/dd/GateMatrixDefinitions.hpp:1]: Hm. If the only production consumer of the DD Package is the DDSIM QDMI device, then I would argue that we can also require MLIR for the package builds.
MLIR is the only way to build program representations that can be transformed to DDs.
I would argue that standalone usage of the package without that is borderline useless.
As such, we may even take this one step further, and directly integrate the DD Package into the MLIR subtree of the repository.
However, in order to avoid scope creep and to get this PR merged asap, let's also open up a separate tracking issue for this and see whether we can clean this up before the `v4.0` release.

File: python/mqt/core/plugins/qiskit/backend.py (hunk 1 above)
Comment (burgholzer) [reply python/mqt/core/plugins/qiskit/backend.py]: It would also be fine for me to move this to a dedicated follow-up issue in the interest of getting this PR merged.

File: include/mqt-core/dd/GateMatrixDefinitions.hpp
Comment (simon1hofmann) [reply include/mqt-core/dd/GateMatrixDefinitions.hpp:1]: 🤖 This response was generated with AI assistance.

Opened #2331 to track consolidating the DD package with the MLIR stack, including the build boundary and duplicate gate-matrix definitions. I kept the architectural work out of #2288.

File: python/mqt/core/plugins/qiskit/backend.py (hunk 1 above)
Comment (simon1hofmann) [reply python/mqt/core/plugins/qiskit/backend.py]: 🤖 This response was generated with AI assistance.

Opened #2332 to track a register-wide initialization path and the corresponding Qiskit serializer cleanup. The issue explicitly includes resolving the OpenQASM typing-versus-Qiskit-compatibility question before implementation.

**Review by burgholzer — APPROVED** [review]
Approving this now to get it in once CI is green! 🚀 
Issues for all kinds of follow-ups have been created.
So let's ship this :shipit:

**Comment by simon1hofmann** [comment]
Seems like the maximum diff for linting is 20000 lines 🤔

**Comment by burgholzer** [comment]
> Seems like the maximum diff for linting is 20000 lines 🤔 

Yeah. We can simply ignore the failing check here and simply merge if everything else is green.

**Comment by simon1hofmann** [comment]
> > Seems like the maximum diff for linting is 20000 lines 🤔 
> 
> Yeah. We can simply ignore the failing check here and simply merge if everything else is green.

Would be ready to merge now.

**Comment by burgholzer** [comment]
Let's go 🚀
